operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/internal/InformerEventSource.java

```diff
@@ -0,0 +1,85 @@
+package io.javaoperatorsdk.operator.processing.event.internal;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+import io.fabric8.kubernetes.client.informers.SharedInformer;
+import io.fabric8.kubernetes.client.informers.cache.Store;
+import io.javaoperatorsdk.operator.processing.event.AbstractEventSource;
+
+public class InformerEventSource<T extends HasMetadata> extends AbstractEventSource {
+
+  private final SharedInformer<T> sharedInformer;
+  private final ResourceToRelatedCustomResourceUIDMapper<T> mapper;
+  private final boolean skipUpdateEventPropagationIfNoChange;
+
+  public InformerEventSource(SharedInformer<T> sharedInformer,
+      ResourceToRelatedCustomResourceUIDMapper<T> mapper) {
```

Comment:
Wonder if we should use `Predicate<CustomResource>` instead of adding another mapping

Reply:
Not sure how that would work, we really need to produce an UID here. So we can associate the event with the related custom resource.

Reply:
My understanding is that the `InformerEvent` that is generated by the event source relies on the `Predicate<CustomResource>` behind the scenes as when it is constructed with an UID it invokes `new UIDMatchingPredicate(relatedCustomResourceUid)` so I guess it could be made generic and just pass the `Predicate<CustomResource>` from the `InformerEventSource` constructor.

Reply:
I'm checking this now, and discovered one thing, how this predicate is used:
```
  public Set<String> getLatestResourcesUids(Predicate<CustomResource> selector) {
    try {
      lock.lock();
      return resources.values().stream()
          .filter(selector)
          .map(r -> r.getMetadata().getUid())
          .collect(Collectors.toSet());
    } finally {
      lock.unlock();
    }
  }
```
But this is quite un-efficient, before it was just getting the resource from a hashmap using an the UID. Now this is traversing a possible hube map.

Reply:
We will have to change this anyways in V2. I mean how we are addressing the resources (will have to be namespace+name). So, hmm, not sure, for now I would rather stick with the current implementation, so that change is not even more confusing in the future.

Reply:
Yes we need to improve this area i.e. 

I think if we use and indexed informer, then we can improve the performances here. 

The issue withe the UID is that is not always possible to use it, in my case as example a change to a monitored resource could lead to a reconcile loop to be triggered on a number of custom resources thus the need to have something like that.

I guess in the future this selector could be defined with something like: 

```java
List<CustomResource> apply(ReadOnlyStore store);
```

So then the selector can implement an optimized version if the resource can be addressed by a key.

Reply:
The problem with this that, the custom resource itself is not accessible at this moment. So this as in the example applies for a config map, we just know the owner reference or something what we explicitly add to the resource (UID for example).

Also why would it be a list?

In v2 the referencing of the CustomResource will  be always namespace+name (known already known). So reading this from owner reference would be possible.

Reply:
The issues is that you may want to watch resources that are not owned by you (i.e. created by another operator) so you can't have a 1-1 relation and you can't rely on the owner ref but instead you add information to your custom resources about i.e. the name of the external resources.

In such case, when you receive an event related to external resource, you need to scan the operator cache and find any custom resource for which the metadata matches and since there no guarantee that the relation is always 1-1, then you need to return a list.

Reply:
That is a good point. But in that case we have to make the cache available for the custom resource event source. What makes these changes much more complex. Especially, since shortly we will make the `CustomResourceEventSource` based on informer. Where then we will have completely different API.

Reply:
> The issues is that you may want to watch resources that are not owned by you (i.e. created by another operator) so you can't have a 1-1 relation and you can't rely on the owner ref but instead you add information to your custom resources about i.e. the name of the external resources.

On the other hand don't think this applies within one controller. So the that a resource is a dependent resource of multiple custom resources is probably not the case. Although it can happen that a resource change can have impact to multiple custom resource.

Reply:
Lots of good point in this discussion, sdk v2.0 will be awesome :)